File: programmers/java/src/level3/greedy/ConnectIslands.java
package level3.greedy;
// https://programmers.co.kr/learn/courses/30/lessons/42861?language=python3

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public class ConnectIslands {
    static class Edge {
        int x;
        int y;
        int c;

        public Edge(int x, int y, int c) {
            this.x = x;
            this.y = y;
            this.c = c;
        }
    }

    private static int[] parents;

    public static void main(String[] args) {
        ConnectIslands connectIslands = new ConnectIslands();
        System.out.println(connectIslands.solution(4, new int[][]{{0, 1, 1}, {0, 2, 2}, {1, 2, 5}, {1, 3, 1}, {2, 3, 8}}));
    }

    public int findP(int c) {
        if (parents[c] != c) {
            parents[c] = findP(parents[c]);
        }
        return parents[c];
    }

    public void union(int x, int y) {
        x = findP(x);
        y = findP(y);
        if (x < y) {
            parents[y] = x;
        } else {
            parents[x] = y;
        }
    }

    public int solution(int n, int[][] costs) {
        parents = new int[n];
        for (int i = 0; i < n; i++) {
            parents[i] = i;
        }

        List<Edge> edges = new ArrayList<>();
        for (int[] cost : costs) {
            edges.add(new Edge(cost[0], cost[1], cost[2]));
        }
        edges.sort(Comparator.comparing(edge -> edge.c));
        int answer = 0;
        for (Edge edge : edges) {
            if (findP(edge.x) != findP(edge.y)) {
                union(edge.x, edge.y);
                answer += edge.c;
            }
        }
        return answer;
    }
}

File: programmers/python/level3/greedy/connect_islands.py
# https://programmers.co.kr/learn/courses/30/lessons/42861?language=python3
def find_p(c):
    if parent[c] != c:
        parent[c] = find_p(parent[c])
    return parent[c]

def union(x, y):
    x = find_p(x)
    y = find_p(y)
    if x < y:
        parent[y] = x
    else:
        parent[x] = y    

def solution(n, costs):
    #섬번호가 순서대로 주어진다고 가정 0 ~ n-1
    global parent
    parent = [0] * (n)
    edges = []

    for i in range(n):
        parent[i] = i
    
    for cost in costs:
        x, y, c = cost
        edges.append((c, x, y))

    edges.sort()

    result = 0

    for edge in edges:
        c, x, y = edge
        if find_p(x) != find_p(y) :
            union(x, y)
            result += c

    return result

print(solution(4, [[0,1,1],[0,2,2],[1,2,5],[1,3,1],[2,3,8]]))
